### ISSUE TITLE
[Disabled] directive bug for form controls

### ISSUE THREAD

**Issue by leandrotassi** [angular/angular#48350]
### Command

other

### Is this a regression?

- [X] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

Angular 14

### Description

When upgrading from Angular 14 to Angular 15 the [Disabled] directive of form controls stopped working.

Is this a bug or did something change?

 I'm using version 15.0.2

### Minimal Reproduction

 <button type="button" id="btnAddPeopleEditor" class="btn btn-success btn-sm" (click)="addPeople(1)"
                  [disabled]="disabledEditEditor || (habilitarControle(isReadOnly) == 'true')">
                  Add
                </button>

### Exception or Error

```text
No error occurs. It simply stopped working.
```

### Your Environment

```text
Angular CLI: 15.0.2
Node: 16.17.0
Package Manager: npm 8.18.0
OS: win32 x64

Angular: 15.0.2
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1500.2
@angular-devkit/build-angular   15.0.2
@angular-devkit/core            15.0.2
@angular-devkit/schematics      15.0.2
@angular/cdk                    15.0.1
@angular/material               15.0.1
@nguniversal/builders           15.0.0
@nguniversal/common             15.0.0
@nguniversal/express-engine     15.0.0
@schematics/angular             15.0.2
rxjs                            7.5.7
typescript                      4.8.4
```

**Comment by jnizet** [comment]
Hi @leandrotassi 

I can't reproduce: https://stackblitz.com/edit/angular-ivy-dqrsbv?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html

It's most probably a bug in your code. You should ask on stackoverflow or on the Angular discord, and provide a complete minimal reproduction.

**Comment by leandrotassi** [comment]
Hi @jnizet 

the bug occurs when using the formControlName within the formGroup.

You can look in the 'profile-editor.component.html' component and compare two controls:

1. `<input [disabled]="true"  id="street"  type="text"  formControlName="street" />`

Does not work.

2. 

`<input [disabled]="true" id="text_disable1" type="text"/>`

It works

Here's an example:

[https://stackblitz.com/edit/angular-cu6mce?file=src/app/profile-editor/profile-editor.component.html](url)

**Comment by jnizet** [comment]
You're not supposed to use disabled when binding to a form control. Look at the browser  console: a message is telling you that. See https://stackblitz.com/edit/angular-ivy-dqrsbv?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.module.ts,src%2Fapp%2Fapp.component.html for example. Open the console, and you'll see a big warning message saying:

> It looks like you're using the disabled attribute with a reactive form directive. If you set disabled to true
> when you set up this control in your component class, the disabled attribute will actually be set in the DOM for
> you. We recommend using this approach to avoid 'changed after checked' errors.

>  Example:
> // Specify the `disabled` property at control creation time:
> form = new FormGroup({
>   first: new FormControl({value: 'Nancy', disabled: true}, Validators.required),
>   last: new FormControl('Drew', Validators.required)
> });

> // Controls can also be enabled/disabled after creation:
> form.get('first')?.enable();
>  form.get('last')?.disable();

And that warning exists for a long time, well before Angular 15.

**Comment by CynanX** [comment]
The `attr.diabled` direction has also stopped working, to which there has been no warning in Angular about not using - is this the same, in that it is not supposed to be supported by Angular 15 now? Thanks.

**Comment by leandrotassi** [comment]
That [Disabled] directive actually stopped working in version 15 and I don't even get the warning in the browser console.
It was working perfectly in version 14.
But, anyway, I changed the approach to disable the controls at form creation time, as recommended by Angular.

**Comment by e-oz** [comment]
[attr.disabled] and [disabled] doesn't add "disabled" attribute anymore.
Quite important change, actually, please mention such changes as "breaking".

**Comment by e-oz** [comment]
>  <input [disabled]="true" id="text_disable1" type="text"/>
>  It works

Doesn't work in `15.1.0-next.3` - it only adds `ng-reflect-is-disabled="true"` to the element in DOM.

**Comment by e-oz** [comment]
> You're not supposed to use disabled when binding to a form control

@jnizet calling `control.disable()` doesn't work either.

**Comment by e-oz** [comment]
Here is StackBlitz with reproduction:
https://stackblitz.com/edit/angular-ivy-jcwird?file=src/app/app.component.html

**Comment by leandrotassi** [comment]
> Here is StackBlitz with reproduction: https://stackblitz.com/edit/angular-ivy-jcwird?file=src/app/app.component.html

Thanks.  At the moment I have no way to test it.  As soon as it works, I return the result.

**Comment by leandrotassi** [comment]
> I'm trying to reproduce the regression so I can figure out why this changed. However, it seems to behave exactly the same (odd) way in Angular 14: https://stackblitz.com/edit/angular-ivy-4ymuur?file=src/app/app.component.html
> 
> @leandrotassi , can you help me show it working in Angular 14?

Thanks.  At the moment I have no way to test it.  As soon as it works, I return the result.

**Comment by leandrotassi** [comment]
> I'm trying to reproduce the regression so I can figure out why this changed. However, it seems to behave exactly the same (odd) way in Angular 14: https://stackblitz.com/edit/angular-ivy-4ymuur?file=src/app/app.component.html
> 
> @leandrotassi , can you help me show it working in Angular 14?

@dylhunn 

Two examples follow:

An input control using [attr.disabled] in reactive form that works in Angular 14.

[](https://stackblitz.com/edit/angular-ivy-ecordk?file=src/app/app.component.html)

An input control using [attr.disabled] in reactive form not working in Angular 15.

[](https://stackblitz.com/edit/angular-ivy-pccxpj?file=src/app/app.component.ts)

**Comment by leandrotassi** [comment]
> > I'm trying to reproduce the regression so I can figure out why this changed. However, it seems to behave exactly the same (odd) way in Angular 14: https://stackblitz.com/edit/angular-ivy-4ymuur?file=src/app/app.component.html
> > @leandrotassi , can you help me show it working in Angular 14?
> 
> Thanks. At the moment I have no way to test it. As soon as it works, I return the result.

@e-oz 

In this example using this version, the alert does not appear in the console

[https://stackblitz.com/edit/angular-ivy-3dt6jt?file=src/app/app.component.html](url)

**Comment by dylhunn** [comment]
> An input control using [attr.disabled] in reactive form that works in Angular 14.
> 
> https://stackblitz.com/edit/angular-ivy-ecordk?file=src/app/app.component.html
>
> An input control using [attr.disabled] in reactive form not working in Angular 15.
>
> https://stackblitz.com/edit/angular-ivy-pccxpj?file=src/app/app.component.ts

Thanks for the reproduction @leandrotassi. This is a serious issue that is likely to break a lot of forms in many apps.

The only PR touching the `formControlName` directive in v15 was #43499 by @crisbeto. This could be the cause -- I'll have to write a test to check.

**Comment by e-oz** [comment]
@leandrotassi your link is broken and alert in console was intentional (I wonder how you missed my warning about it).

**Comment by e-oz** [comment]
@dylhunn check my repro also, please - [attr.disabled] and control.disable() - they are affected also

**Comment by ToomeyDamien** [comment]
@e-oz 

`control.disable()` works in your demo right? (https://github.com/angular/angular/issues/48350#issuecomment-1362793387)

Doing

```ts
constructor() {
  this.formGroup.disable()
}
```

or

```ts
constructor() {
  this.formGroup.get('field').disable()
}
```

disables the radio buttons as expected on my side.

**Comment by e-oz** [comment]
@ToomeyDamien you are right, the issue (in this particular case) was in that custom directive. I've [fixed it](https://stackblitz.com/edit/angular-ivy-pp19ra?file=src%2Fapp%2Fdisable-control.directive.ts,src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html) and removed from [my demo](https://stackblitz.com/edit/angular-ivy-jcwird?file=src%2Fapp%2Fapp.component.html,src%2Fapp%2Fapp.module.ts,src%2Fapp%2Fapp.component.ts).

**Comment by juanmendes** [comment]
Since you should not use `attr.disabled` when using `formControlName`, I think the issue I created has a simpler reproduction of the problem. Please see https://github.com/angular/angular/issues/48737

I'm implementing my own `ControlValueAccessor` which is why I bind to `attr.disabled`. Simple reproduction:


```ts
@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule, FormsModule],
  template: `<input [attr.disabled]="true"  [(ngModel)]="myName" /> `,
})
export class AppComponent {
  myName = 'Angular';
}
```

**Comment by JeanMeche** [comment]
About the `attr.disabled` "bug", this a [breaking change/feature](https://angular.io/guide/update-to-version-15#setdisabledstate-is-always-called-when-a-controlvalueaccessor-is-attached), the "fix" is to opt-out from the new behaviour by importing the `ReactiveFormsModule` with a config.

```
ReactiveFormsModule.withConfig({
      callSetDisabledState: 'whenDisabledForLegacyCode',
    }),
```

**Comment by JeanMeche** [comment]
> The only PR touching the `formControlName` directive in v15 was #43499 by @crisbeto. This could be the cause -- I'll have to write a test to check.

@dylhunn, I can conform this broke the behaviour : calling `this.valueAccessor!.setDisabledState!(true);` was what disabled the input.

**Comment by aleesaan** [comment]
I think that [this PR](https://github.com/angular/angular/pull/47576) actually introduced a regression @dylhunn, as now I cannot disable a single radio button anymore - the only way is to use the old behavior via `callSetDisabledState: 'whenDisabledForLegacyCode'`.  
Take [the example in the RadioControlValueAccessor docs](https://angular.io/api/forms/RadioControlValueAccessor#using-radio-buttons-with-reactive-form-directives), how could we disable individual radio buttons? Before I used to do it via `[attr.disabled]`, but now that doesn't work anymore of course. Disabling the whole control is not an option, as it disables all the radios, so how would we (with this new way) disable radio inputs individually? It's a valid use case, right?

**Comment by argelj289** [comment]
I am having some issue for formcontrol wherein the disable attribute is not attaching to the angular html template. any reason for this? this is for angular 15 version

**Comment by dylhunn** [comment]
Hi everyone. Indeed, I've tracked this down, and @JeanMeche is correct. While trying to fix a bug in https://github.com/angular/angular/pull/47576 I accidentally caused this breakage. For now, you can work around by opting out when you import forms:

```
ReactiveFormsModule.withConfig({
    callSetDisabledState: 'whenDisabledForLegacyCode',
}),
```

or, if you're using template-driven forms:

```
FormsModule.withConfig({
    callSetDisabledState: 'whenDisabledForLegacyCode',
}),
```

I expect to have this fixed for real in a patch release soon.

**Comment by dylhunn** [comment]
After considerable debugging, I got to the bottom of this.

## tl;dr

- Using Reactive Forms? You should be setting `disabled` on your model, not your template. Try `new FormControl({value: 'foo', disabled: true})`. (Or you could call `myControl.disable()` in `ngOnInit`.)
- Want to opt-out of the fix? Make sure you're on `15.1.0` or later and import `FormsModule.withConfig({callSetDisabledState: 'whenDisabledForLegacyCode'})` (or `ReactiveFormsModule`, if that's what you're using).
- Using `[attr.disabled]` on radio buttons? This is currently broken, and will be fixed in #48864.

## Full Explanation

`setDisabledState` is supposed to be called whenever the disabled state of a control changes, including upon control creation. However, a longstanding bug caused the method to not fire when an *enabled* control was attached. This bug was fixed in v15, in #47576.

This had a side effect: previously, it was possible to instantiate a reactive form control with `[attr.disabled]=true`, even though the the corresponding control was enabled in the model. (Note that the similar-looking property binding version `[disabled]=true` was always rejected.) This resulted in a mismatch between the model and the DOM -- the model would think the control was enabled, when it was actually disabled. Now, because `setDisabledState` is always called, the value in the DOM will be immediately overwritten with the "correct" enabled value.

Users of Reactive Forms should instead disable the control directly in their model. (There are many ways to do this, such as using the `{value: 'foo', disabled: true}` constructor format, or immediately calling `FooControl.disable()` in `ngOnInit`.)

If this incompatibility is too breaking, you may also opt out using `FormsModule.withConfig` or `ReactiveFormsModule.withConfig` at the time you import the module, via the `callSetDisabledState` option.

However, there is an exceptional case: radio buttons. Because Reactive Forms models the entire group of radio buttons as a single `FormControl`, there is no way to control the disabled state for individual radios from the model. (This is probably a design oversight in the forms package!) In #48864, we have special cased radio buttons to ignore their first call to `setDisabledState` when in `callSetDisabledState: 'always'` mode. This preserves the old behavior, allowing `[attr.disabled]` to keep working.

**Comment by e-oz** [comment]
> Using Reactive Forms? You should be setting disabled on your model, not your template.

We can declare reactive form attributes in the template, but only some of them - that's not cool.
It is not always possible (or brings a lot of code) to modify some attributes based on input values - because they might change. When it's in the template, Angular change detection will monitor the changes and update bindings, when it's in the code - you have to do this, and it's not always trivial.

Here is a directive as a workaround:
```ts
import { NgControl } from '@angular/forms';
import { ChangeDetectorRef, Directive, Input, OnChanges, OnInit } from '@angular/core';

@Directive({
  selector: '[disableControl]',
  standalone: true,
})
export class DisableControlDirective implements OnInit, OnChanges {
  constructor(private ngControl: NgControl, private cdr: ChangeDetectorRef) {
  }

  @Input() disableControl?: boolean | null;

  ngOnInit() {
    this.update(!!this.disableControl);
  }

  ngOnChanges() {
    this.update(!!this.disableControl);
  }

  private update(disable: boolean) {
    try {
      const ctrl = this.ngControl.control;
      if (ctrl) {
        if (disable) {
          if (!ctrl.disabled) {
            ctrl.disable({onlySelf: true, emitEvent: false});
            this.cdr.markForCheck();
          }
        } else {
          if (ctrl.disabled) {
            ctrl.enable({onlySelf: true, emitEvent: false});
            this.cdr.markForCheck();
          }
        }
      }
    } catch (e) {
      console?.error(e);
    }
  }
}
```

**Comment by benartmon** [comment]
@dylhunn Thanks for this detailed information. A question though: if a project introduces `.withConfig({callSetDisabledState: 'whenDisabledForLegacyCode'})` into the code, is this something that will be deprecated anytime soon? What would be the path to get rid of it?

**Comment by dylhunn** [comment]
@benartmon Far too many projects rely on the old behavior, including thousand of tests inside of Google. We may or may not deprecate it at some distant future point, but the option will never be removed.

**Comment by kevinbuhmann** [comment]
> Using Reactive Forms? You should be setting disabled on your model, not your template.

I want to echo @e-oz here. This is a good solution in my opinion. We can set reactive form validators in our template using native attributes like `required`, `min`, `max, etc. Why should `disabled` be different?

**Comment by JensWouters** [comment]
@dylhunn In the original question a button was used, but all the answers look like they are using an input with a FormControl, which a button can't use. 

I have a component in which a button is used with the disabled attribute and this no longer works. 
If I use the withConfig fix in the module where my module of the component is imported it doesn't fix it.

Any idea on a solution for disabled attribute in buttons?

**Comment by rbirkgit** [comment]
We encountered this issue to when setting `disabled` attribute on template driven select/input[checkbox]. 

We have dynamic access rights and in ngAfterViewInit() we set the attribute when in read-only mode. It worked in ang14, but with ang15 is stopped in our template driven forms. Our reactive ones worked fine. This is with latest 15.2.8

The `disabled` attribute never get set. It's like the angular code removes it. Our directive sets it like this;
```
          renderer.setAttribute(el, 'disabled', '');
```

The only way to get it working again was to disable the new disable functionality with:
```
FormsModule.withConfig({ callSetDisabledState: 'whenDisabledForLegacyCode' }),
```

**Comment by PhiLhoSoft** [comment]
I have a different use case, and I confirm that when we call `formControl.disable()`, the `setDisabledState` function is no longer called.

We don't use `[disabled]` nor `[attr.disabled]` in our code, but the change affected us, in a way that doesn't seem right.
The `.withConfig({ callSetDisabledState: 'whenDisabledForLegacyCode' })` works, but it means we have to do in on each of our declaration, which is ugly and inconvenient.

The use case, quite common: we have custom form components, for checkbox, radio, text field and so on.
These components declare `NG_VALUE_ACCESSOR` and implement `ControlValueAccessor` and use `setDisabledState` to set the `disabled` attribute on the native element. I know there are better ways, but that's legacy code that worked well for years.
If we disable the form group where the controls are, it works. If we disable individual components, it no longer works.

**Comment by sandeep3333** [comment]
> Hi @jnizet
> 
> the bug occurs when using the formControlName within the formGroup.
> 
> You can look in the 'profile-editor.component.html' component and compare two controls:
> 
> 1. `<input [disabled]="true"  id="street"  type="text"  formControlName="street" />`
> 
> Does not work.
> 
> 
> `<input [disabled]="true" id="text_disable1" type="text"/>`
> 
> It works
> 
> Here's an example:
> 
> [https://stackblitz.com/edit/angular-cu6mce?file=src/app/profile-editor/profile-editor.component.html](url)

I have the same issue so for the workaround i have used [readonly]

**Comment by polonmedia** [comment]
@dylhunn or anybody, can you help me with radio control value accessor?
Im trying to implement this:
![image](https://github.com/angular/angular/assets/59230634/dd75ef9d-b507-4715-bac1-4b8cb280fc5e)
and I get following errors:
`error TS2322: Type 'ModuleWithProviders<FormsModule>' is not assignable to type 'readonly any[] | Type<any>'.`
and
`'imports' contains a ModuleWithProviders value, likely the result of a 'Module.forRoot()'-style call. These calls are not used to configure components and are not valid in standalone component imports - consider importing them in the application bootstrap instead.(-992012)`

how to get FormsModule.withConfig in standalone component? Angular ver 16.2.0

**Comment by jnizet** [comment]
@polonmedia 
```
bootstrapApplication(AppComponent, { providers: [importProvidersFrom(FormsModule.withConfig(...), ...)
```

**Comment by polonmedia** [comment]
@jnizet thank u for fast reponse man, youre awesome
using "whenDisabledForLegacyCode" now and all of the radio buttons are disabled, the warning still persists
![image](https://github.com/angular/angular/assets/59230634/1507d9e2-05c1-4d75-b250-216069cc3a98)

with `[attr.disabled]="true"` warning disappears but still all of the inputs are disabled

is there really no proper way to set [disabled] in html, no proper way to disable radio via formControl or at least get rid of the warning somehow so it falls back to legacy? seems weird

**Comment by johncrim** [comment]
This is also a problem, in that a warning spews to the console if you use any of the `formControl` directives on a component, and the component has an `@Input() disabled` . IMO the benefit of the warning in a very specific case is not worth the noise.

To state it another way: I don't have a problem with `disabled` not working with `formControl` directives. I have a problem that I'm getting an unnecessary warning (in dev builds) for using an `@Input() disabled` on my components, if I use a `formControl` directive on those components. The name `disabled` is very useful and clear, and I think it's unreasonable to assume that it can only be used on the `formControl` directives. 

Eg:
```ts
@Component({
  selector: 'ui-listbox',
  ...
  providers: [ {
    provide: NG_VALUE_ACCESSOR,
    useClass: SelectionValueAccessor,
    multi: true
  } ]
})
export class ListboxComponent<TKey> {

  @Input()
  set disabled(v: boolean) {
    this.focusArea.disabled = v;
    this.state.disabled = v;
  }
  get disabled() {
    return this.state.disabled;
  }
  
  ...
}

// Use in template:
<ui-listbox [disabled]="disabled"
      [formControl]="formControl">
```

gives me this warning even though nothing is wrong. Please remove the warning.

**Comment by johncrim** [comment]
I should add: `formControl.disabled` has different meaning from a typical component's `MyComponent.disabled`. When a `FormControl` is disabled, it means "exclude the values from this FormControl in the parent FormGroup values". For a web page element, or UI component, a `disabled` attribute typically means "appears in accessible, cannot be focused, and value cannot be changed by user".

I think it makes sense that `FormControlDirective.disabled` doesn't set `FormControl.disabled`. But it's problematic to have this erroneous warning, where the options are either to a) ignore a bunch of console warnings, or b) rename `disabled` to `realDisabled` or something equally irritating.

**Comment by EliseoPlunker** [comment]
As temporaly solution create a directive
```
@Directive({
  selector: '[disable]',
  standalone:true
})
export class CustomDisabledDirective implements AfterViewInit {
  _disable:boolean
  @Input('disable') set _(value:boolean)
  {
    this._disable=value;
    this.ngAfterViewInit();
    
  }
  constructor(private elementRef:ElementRef) { }
  ngAfterViewInit()
  {
    if (this._disable)
    this.elementRef.nativeElement.setAttribute('disabled','true')
    else
    this.elementRef.nativeElement.removeAttribute('disabled')
  }

```

**Comment by pepijnmm** [comment]
For use [disabled] already stopped working with angular 13 when it had a formControlName. We had to use [attr.disabled] from then on.

It seems it was mentioned in the breaking changes but I agree that it could have been worded better [release note](https://angular.io/guide/update-to-version-15#setdisabledstate-is-always-called-when-a-controlvalueaccessor-is-attached)

We don't want the input to be disabled in the formGroup as validation still needs to be checked as the input gets filled in based on the other inputs from the user.
So our solution is to change in our app.module.ts `ReactiveFormModule` to `ReactiveFormsModule.withConfig({callSetDisabledState: 'whenDisabledForLegacyCode'})`

[dylhunn](https://github.com/dylhunn) explains it better in his comment: https://github.com/angular/angular/issues/48350#issuecomment-1405900515
It would be nice if attr.disabled could be checked before it sets disabled on false again. but as far as I could tell that attribute was already removed and only disabled was still on the nativeElement.

**Comment by imasip** [comment]
left working from v.13 to v.17 but works when move control outside the form.

**Comment by Sanjeev-92** [comment]
i am using common component to render form after upgrade to angular 16 [attr.disabled] is not working without throwing any error.

its a huge blunder in angular upgrade

**Comment by mahalinge** [comment]
am using [attr.disabled] in multiple comp template to disable based on conditions ,after upgrading to ang17 ,its stopped working ,any work around ?

**Comment by mahalinge** [comment]
workaround we can add custom directive and use it in template input wherever it needed.

**Comment by ilyakonrad** [comment]
I'm trying to use conditional `disabled` on `mat-option` but it only works sometimes.
I think it's due to the fact that I also disable/enable the control programmatically.

It doesn't make sense. What is the "right" way to do it in reactive forms? 🤦

**Comment by RAHMANIABOUBEKER** [comment]
Hello,

Dose this concern all form field component or just checkbox ?

**Comment by Laffery** [comment]
Any progress?

**Comment by mrmokwa** [comment]
Just want to give some feedback of my needs.

I want to be able to disable my form control and display the form control errors at the same time.

**Comment by Stevetanus** [comment]
The workaround from @EliseoPlunker 
> ```
> export class CustomDisabledDirective implements AfterViewInit {
>   _disable:boolean
>   @Input('disable') set _(value:boolean)
>   {
>     this._disable=value;
>     this.ngAfterViewInit();
>     
>   }
>   constructor(private elementRef:ElementRef) { }
>   ngAfterViewInit()
>   {
>     if (this._disable)
>     this.elementRef.nativeElement.setAttribute('disabled','true')
>     else
>     this.elementRef.nativeElement.removeAttribute('disabled')
>   }
> ```

Then I did a little improvement as below using `Renderer2`
```
import { AfterViewInit, Directive, ElementRef, Input, Renderer2 } from '@angular/core';

@Directive({
  selector: '[disabledDirective]'
})
export class CustomDisabledDirective implements AfterViewInit{
  _disabled: boolean;
  @Input() set disabledDirective(value: boolean) {
    this._disabled = value;
    this.updateDisabledState();
  }

  constructor(private el: ElementRef, private renderer: Renderer2) { }

  ngAfterViewInit(): void {
    this.updateDisabledState();
  }

  updateDisabledState(): void {
    if (this._disabled) {
      this.renderer.setAttribute(this.el.nativeElement, 'disabled', '');
    } else {
      this.renderer.removeAttribute(this.el.nativeElement, 'disabled');
    }
  }
}
```
